(mixedstates)=

# Смешанные состояния и матрицы плотности

## Описание лекции

Из этой лекции мы узнаем:

- Что такое матрица плотности
- Как ввести в описание квантовых состояний новый уровень случайности
- Как связаны смешанные (mixed) и запутанные (entangled) состояния
- Как можно описать насколько "сильно" квантовое состояние является смешанным

## Смешанные состояния

Эта глава кратко рассматривает довольно сложный, но крайне важный аспект квантового описания мира: как описать квантовую систему, в которой случайным является не только результат измерения, но и само состояние квантовой системы, иными словами, когда мы не можем описать квантовую систему определенным вектором состояния.

Как могут появиться такие ситуации?

Прежде всего, в любой реалистичной ситуации квантовая система будет испытывать влияние окружающей среды. Поскольку уровень тепловых шумов при обычной ("комнатной") температуре достаточно высокий, то симуляция реалистических квантовых систем должна учитывать влияние шумов. Во многих ситуациях такие шумы приводят к смешанным состояниям.

Вторая, более важная причина использования смешанных состояний состоит в их тесной связи с запутанными состояниями, как показано ниже.
## Почему нельзя обойтись волновой функцией

Рассмотрим реальный физический кубит в состоянии равновесия, например спин ядра атома в магнитном поле.
Статистическая физика говорит нам, что вероятность обнаружить этот спин ориентированным вдоль внешнего поля выше, чем в
противоположном полю направлении. Для расчёта отношений этих вероятностей можно использовать распределение Больцмана:
$p_\uparrow/p_\downarrow = e^{\frac{2\mu \Delta H}{k_\mathrm{B}T}}$, где $2\mu \Delta H$ -- разность энергий остояний со спином вдоль
поля и в протовоположном полю направлений, $k_B$ -- постоянная Больцмана, а $T$ -- температура. Так как состояний у этой
системы всего два, сумма их вероятностей должна давать единицу: $p_\uparrow + p_\downarrow = 1$. Таким образом,
вероятности состояний однозначно определены.

Волновую функцую, отвечающую такому тепловому состоянию, можно записать в виде

$$
\ket{\psi} =
\begin{pmatrix}
\sqrt{p_\uparrow} \\
\sqrt{p_\downarrow}e^{i\phi}
\end{pmatrix},
$$

где значение фазы $\phi$ не определено. Однако $\phi$ определяет поведение системы не в меньшей мере, чем вероятности
$p_\uparrow$ и $p_\downarrow$. $\phi$ может равновероятно принимать любые значения. Таким образом, уже даже состояние
теплового расвновесия нельзя описать одной волновой функцией -- это будет распределением вероятности по волновым
функциям с разным $\phi$.

## Матрица плотности

Оказывается, что вместо распределения вероятностей по волновым функциям можно использовать более простую
конструкцию -- матрицу плотности. Если система находится в состояниях $\Phi_n$ с веротяностями $p_n$, то матрицу
плотности можно определить как

$$
\rho = \sum\limits_{n}p_n \ket{\Phi_n}\bra{\Psi_n}.
$$ (eqn:rho_mixed)

Выражение $\ket{\Phi_n}\bra{\Phi_n}$ обозначает произведение вектора-столбца на вектор строку -- результатом будет
матрица. Важно, что значение любой ожидаемой величины (отвечающей оператору $\hat{A}$) можно записать через $\rho$:

$$
\mathbb{E}[A] = \sum\limits_{n}p_n \bra{\Phi_n}\hat{A}\ket{\Phi_n} =
\sum\limits_{n}p_n \operatorname{Tr}\left[\ket{\Phi_n}\bra{\Phi_n}\hat{A}\right] = \operatorname{Tr}\left[\rho \hat{A}\right]
$$

Математическое обоснование этой циклической перестановки можно получить, используя покомпонентную поэлементную сумму
для произведений:

$$
\bra{\Phi_n}\hat{A}\ket{\Phi_n} = \sum \limits_{i,j} \Phi_n^{i*} A_{ij} \Phi_n^{j} =
\operatorname{Tr}\left[\ket{\Phi_n}\bra{\Phi_n}\hat{A}\right].
$$


Рассмотрим ситуацию, когда описываемую квантовую систему $\Phi$ можно разделить на две подсистемы, $\phi$ и $\psi$, и состояние этой системы $\ket{\Phi}$ является суперпозицией состояний двух подсистем:

$$
\ket{\Phi} = \sum_{i,j} c_{i,j} \ket{\phi_i} \ket{\psi_j}
$$

(здесь два ортонормированных набора векторов состояния $\ket{\phi_i}$ и $\ket{\psi_j}$ описывают две части всей системы).

Для такого состояния нельзя сказать, в каком именно состоянии находится каждая подсистема.

```{note}
Вспомним обсуждение кота Шредингера в [лекции про кубит](qubit) -- упрощяая до предела, можно считать, что радиоактивный атом является одной системой, а несчастный кот -- второй.

До измерения ни атом ни кот не имеют определенного состояния, а находятся в суперпозиции возможных состояний.
```

Что же мы можем сделать, если нужно определить ожидаемое значение какой-то величины, которая имеет смысл только для одной из подсистем (например, распался ли атом)? Такая величина будет описываться оператором $A_\phi$, который действует _только_ на первую подсистему $\phi$ (в примере кота Шредингера будет бесполезно измерять, распался ли кот или жив ли атом). Ожидаемое значение такой физической величины можно посчитать, используя понятие _редуцированной_ матрицы плотности (усреднение общей матрицы плотности по состояниям всей системы $\Phi$)

$$
\rho_\phi = Tr_\psi (\ket{\Phi} \bra{\Phi}) = \sum_{i, j} d_{i, j} \ket{\phi_i} \bra{\phi_j},
$$

где

$$
d_{i, j} = \sum_k a_{i, k} a^{*}_{j, k},
$$

где $Tr_\psi$ означает частичный след по второй подсистеме, а звездочка -- комплексное сопряжение. В таких обозначениях для среднего от оператора $A$ вычисляется по формуле

$$
\braket{A_\phi} = Tr_\phi (A \rho_\phi)
$$

(здесь след матрицы уже вычисляется по первой подсистеме).

Для системы, которая может быть описана одной волновой функцией $\Psi$ выражение для матрицы плотности получается тривиальным:

$$
\rho = \ket{\Psi}\bra{\Psi}
$$ (eqn:rho_pure)

Такие состояния, которые нельзя описать одним вектором состояния, а можно лишь матрицей плотности называются _смешанными_ (англ. _mixed states_), в отличие от _чистых_ состояний (англ. _pure states_), которые описываются вектором состояния или матрицей плотности вида {eq}`eqn:rho_pure`. То есть в данном случае мы не "видим" всей системы, а "видим" лишь какую-то ее часть, которая является вероятностной суммой чистых ее состояний по состояниям той части, которая нам "не видима".

Стоит заметить, что матрица плотности для подсистем была впервые введена в научный оборот знамениты советским физиком, лауреатом Нобелевской премии Львом Ландау {cite}`howtounderstandqm`

```{figure} /_static/qcblock/mixedstates/Landau.jpg
:name: landau

Лев Ландау, 1908-1968
```

Можно ли математически определить разницу между чистыми и смешанными состояниями?

Как можно легко показать, для оператора-проектора {eq}`eqn:rho_pure` выполняется тождество

$$
\rho^2 = \ket{\phi} \bra{\phi} \ket{\phi} \bra{\phi} = \ket{\phi} \bra{\phi} = \rho
$$

и следовательно, для чистого состояния

$$
Tr (\rho^2) = 1
$$

(напомним, что мы всегда считаем вектора состояния нормированными на единицу).

Аналогичным образом, но после более длинных выкладок можно показать, что в общем случае

$$
Tr (\rho^2) \leq 1,
$$

причем знак равенства в последней формуле возможен, только если в формуле {eq}`eqn:rho_mixed` сумма имеет лишь одно слагаемое (т.е. состояние является чистым). Последнее свойство матрицы плотности позволяет ввести ряд величин, характеризующих смешанные и запутанные состояния, как будет рассказано в заключительном разделе этой лекции. Величина $Tr (\rho^2)$ называется чистотой состояния, или, на английском -- quantum state purity.

В заключение этого раздела заметим, что чаще всего при обсуждении смешанных состояний рассматривают только одну "подсистему", считая, что вторая -- это некоторый макроскопический объект ("резервуар", например лаборатория или даже вся Вселенная). В этом случае формулу {eq}`eqn:rho_mixed` можно записать, как

$$
\rho = \sum_i d_i \ket{\phi_i} \bra{\phi_i}.
$$

В таком представлении можно сказать, что смешанное состояние -- это ситуация, когда мы не можем сопоставить квантовой систему никакого конкретного вектора, а считаем, что существует статистическая смесь из несколько возможных состояний $\ket{\phi_i}$, каждое со своей вероятностью $d_i$.

## Запутанные состояния

Давайте вернемся к представлению состояния составной системы и зададимся вопросом: что можно сказать о связи между частями системы с точки зрения квантового описания? Для системы из двух кубитов (двухуровневых систем) такая составная система в общем случае может быть записана в явном виде (в этом разделе мы в основном следуем изложению из книги {cite}`valievkokin`):

$$
\ket{\Phi} = a \ket{0_A 0_B} + b \ket{0_A 1_B} + c \ket{1_A 0_B} + d \ket{1_A 1_B},
$$ (eqn:two_qubit)

где индексы $A$ и $B$ здесь обозначают первый и второй кубиты, соответственно, а условие нормировки дает

$$
\left\langle \Phi \middle| \Phi \right\rangle = |a|^2 + |b|^2 + |c|^2 + |d|^2 = 1.
$$

Теперь, как можно показать, состояние типа {eq}`eqn:two_qubit` может быть представлено в виде произведения состояний двух отдельных кубитов, если $ad = dc$:

$$
\ket{\Phi_s} = \left(a_A \ket{0_A} + b_A \ket{1_A} \right) \otimes \left(a_B \ket{0_B} + b_B \ket{1_B} \right),
$$

где для выражения {eq}`eqn:two_qubit`

$$
\begin{array}{cc} a = a_A a_B & b = a_A b_B \\ c = b_A a_B & d = b_A b_B \end{array}
$$

В других случаях, когда $ad \neq dc$, состояние составной системы не представимо в виде произведения состояний подсистем, такие состояния называют несепарабельными (англ. nonseparable). Другими словами, результат измерения состояния подсистемы $A$ будет зависеть от состояния подсистемы $B$. Это означает, что для квантовых систем возможна нелокальная корреляция. Такое свойство квантовых систем называется _запутыванием_ (англ. entanglement), а сами состояния _запутанными_ (англ. _entangled_).

```{note}
В отличие от анлийского языка, в русском языке не сложилось единой терминологии в отношении запутанных состояний. На момент написания этой лекции (осень 2021 года) в статье [Квантовая запутанность](https://ru.wikipedia.org/wiki/Квантовая_запутанность) русскоязычной Википедии указывается восемь (!) отличающихся терминов для этого явления, например, "запутанность", "перепутанность" или "сцепленность". Сделать с этим что-то сложно, остается только иметь в виду имеющиеся обстоятельства. Мы будем стараться употреблять термин _запутанность_ и, соответственно, _запутанные_ состояния.
```

Приведем пару примеров запутанных состояний:

1. $b = c = 0, a = d = \pm 1/\sqrt{2}$ -- состояние "_шредингеровского кота_", см. {cite}`valievkokin` и {cite}`mogilevtsevkilin`, такая формула для вектора состояния возникает для суперпозиции двух макроскопически различимых состояний одной из подсистем (живой или мертвый кот), например различные конфигурации (моды) излучения лазера.
2. $a = d = 0, b = -c = \pm 1/\sqrt{2}$ -- такое состояние называется ЭПР-парой (EPR, от Einstein-Podolsky-Rosen) и это очень важный пример из истории изучения запутанности в квантовой физике.

```{note}
в 1930-е происходили многочисленные споры об "интерпретации" (сути) квантовой механики. Именно тогда Эйнштейн, Шредингер и их коллеги обратили внимание на несепарабельные состояния и затем Эйнштейном, Подольским и Розеном был сформулирован "парадокс" -- что квантовая механика либо нелокальна (т.е. несовместима с теорией относительности), либо неполна (мы учитываем не все параметры при описании состяния квантовых систем). Именно с дискуссией о сути запутанности связана знаменитая цитата Эйнштейна "Бог не играет в кости." и менее известный ответ Нильса Бора, "Альберт, не указывай Богу, что ему делать."

Позже, уже в конце 20-го -- начале 21-го века, благодаря работам многих других исследователей, см. более полную историю и ссылки на оригинальные статьи например в [Википедии](https://ru.wikipedia.org/wiki/Квантовая_запутанность).
```

Довольно долго изучение запутанности и связанных с ней трудности считались сложным, но неосновными вопросами квантовой физики. Но с развитием квантовой информатики стало понятно, что без запутанности нельзя разрабатывать квантовые компьютеры и системы квантовой связи. В настоящее время существуют устоявшиеся методы создания запутанных состояний в эксперименте. А для целей нашего курса, в симуляциях, достаточно использовать [двухкубитные гейты](../qcblock/gates.html#id15), которые обсуждались в предыдущей лекции, например `CNOT` или `CZ`, который используется в лекции про [Градиенты квантовых схем](gradients).

В качестве примера давайте посмотрим, как можно создать запутанное состояние в PennyLane. Начнем с импортов и создания двухкубитной схемы:

In [1]:
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit", wires=2)

Далее применим к первому кубиту операцию поворота $\hat{RX}$, запутаем кубиты с помощью CNOT и далее оценим запутанность с помощью измерения оператора Паули $\hat{\sigma^z}$:

In [2]:
@qml.qnode(dev)
def circuit(param):
    qml.RX(param, wires=0)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))

В этом примере значение переменной `param` определяет степень запутанности, и для $\pi/2$ запутанности будет максимальна, в результате оба кубита будут максимально смешанными и средний результат измерения будет нулевым:

In [3]:
print(circuit(np.pi / 2))

[2.22044605e-16 2.22044605e-16]


(код взят из официальной [демонстрации](https://pennylane.ai/qml/demos/tutorial_advanced_usage.html) для библиотеки PennyLane)

## Описания эволюции смешанного состояния

### Квантовая динамика

Напомним, что квантовая динамика в терминах волновых функций $\ket{\Psi}$ описывается при помощи уравнения Шредингера:

$$
i\hbar \frac{\partial \Psi(x, t)}{\partial t} = \hat{H} \Psi(x, t)
$$

Аналогичное уравнение можно получить и для матриц плотности. Оно называется уравнением фон-Неймана и записывается через коммутатор $[]$, который определен как $[\hat{A}, \hat{B}] = \hat{A}\hat{B} - \hat{B}\hat{A}$:

$$
i\hbar \frac{\partial \rho}{\partial t} = [\hat{H}, \rho]
$$

### Измерения и томография

Квантовая механика работает так, что любое [измерение](../qcblock/qubit.html#id23) приводит к коллапсу волновой функции и является необратимым. А еще измерение, например, состояния $\ket{+}$ и $\ket{-}$ не различимы при измерениях по оси $\mathbf{Z}$ -- для обоих состояний мы будем получать $\ket{1}$ и $\ket{0}$ с вероятностью $0.5$, то есть нам нужно измерять по _всем_ базисам. В общем, получается, что восстановить амплитуду и фазу волновой функции $\Psi$ это большая проблема, если добавить сюда вероятностный характер измерения.

```{note}
Строго говоря это не просто "большая" проблема, а настоящая NP-полная задача оптимизации!
```

```{figure} /_static/qcblock/mixedstates/MarginalDistribution.png
:width: 250px

Иллюстрация фазовой проблемы.
```

Эта задача обычно решается при помощи квантовой томографии и восстанавливают как раз не волновую функцию $\Psi$, а матрицу плотности $\rho$  (потому что в реальных экспериментах и задачах почти не бывает чистых состояний). Представим, что наша квантовая система описывается базисом $y_i$ -- набором из $2^N$ векторов, причем каждому из этих базисных векторов соответствует свое собственное значение -- результат измерения (подробнее об этом было в [лекции про кубит](../qcblock/qubit.html#id31)). Тогда если у нас будет достаточно много результатов измерений, то мы сможем восстановить нашу матрицу плотности $\rho$ методом максимизации правдоподобия. Выражение для правдоподобия в этом случае можно записать как:

$$
L(\rho) = \prod_i \bra{y_i} \rho \ket{y_i} ^{q_i}
$$

где $q_i$ -- это частота получения собственного значения, соответствующего волновой функции $\ket{y_i}$ (потому что измерение переводит наше состояние в базисный вектор, соответствующий результату измерения). В итоге увеличивая число измерений мы приближаем частоты $q_i$ к вероятностям $\bra{\Psi}(\ket{y_i}\bra{y_i})\ket{\Psi}$, а нашу матрицу $\rho$ к ее истинному виду.

Методы квантовой томографии являются критической частью, в том числе, квантовой связи, так как системы там обычно небольшие, но восстанавливать надо всю матрицу плотности.

## Энтропия

В квантовой механике существует также численная характеристика того, насколько сильно наша система "запутанная". Эта величина называется энтропией фон-Неймана, а ее выражение очень похоже на выражение для классической [энтропии Шеннона](https://ru.wikipedia.org/wiki/Информационная_энтропия). Только в отличии от классики, в квантовой механике мы имеем _матрицу_ плотности, поэтому в выражении у нас фигурирует [матричный логарифм](https://en.wikipedia.org/wiki/Logarithm_of_a_matrix):

$$
S = -Tr(\rho\text{ }ln\text{ }\rho)
$$

В случае "чистого состояния" энтропия также может быть записана через собственные значения матрицы плотности:

$$
S = -Tr(\sum_j e_j \ket{y_j}\bra{y_j}\text{  }ln(\sum_j e_j \ket{y_j}\bra{y_j})) = -\sum_j e_j\text{ }ln\text{  }e_j
$$(rho:eig)

где $e_j$ -- собственное значение, соответствующее собственному вектору $y_j$.

Энтропия фон-Неймана одно из ключевых понятий в квантовой теории, которое играет ключевую роль, например, в изучении черных дыр.

### Сравнение энтропии чистых состояний

Давайте на примере двухкубитной системы посмотрим как меняется энтропия фон-Неймана после применения различных операторов, которые мы обсуждали в [лекции про квантовые гейты](gates). Для начала реализуем функцию, вычисляющую энтропию для заданной матрицы плотности. Так как матричный логарифм считать не всегда удобно мы будем использовать выражение {eq}`rho:eig`. Ну и считать, что $x * \text{  }ln(x) = x * \text{ }ln(x + \epsilon),\text{ }\epsilon\ll 1$.

In [4]:
def entropy(rho: np.array) -> float:
    e = -np.sum(np.diag(rho) * np.log(np.diag(rho) + 1e-16))
    return np.real(e)

```{note}
Тут мы в очередной раз польуземся рутинами из библиотеки линейной алгебры `SciPy`. Документацию по функции `scipy.linalg.logm` можно посмотреть на [официальном сайте](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.logm.html).
```

И реализуем наш гейт $\hat{H}$.

In [5]:
h = 1 / np.sqrt(2) * np.array([
    [1 + 0j, 1 + 0j],
    [1 + 0j, 0j - 1]
])

В качестве начального состояния нашей двухкубитной системы возьмем состояние $\ket{Psi} = \ket{0} \otimes \ket{0}$. И посмотрим энтропию состояний:
- $\ket{\Psi}$
- $(\hat{H} \otimes \hat{I})\ket{\Psi}$
- $(\hat{I} \otimes \hat{H})\ket{\Psi}$
- $(\hat{H} \otimes \hat{H})\ket{\Psi}$

In [6]:
basis_0 = np.array([1 + 0j, 0 + 0j]).reshape((2, 1))
psi = np.kron(basis_0, basis_0)
rho = psi @ psi.conj().T

entropy0 = entropy(rho)
entropy1 = entropy(np.kron(h, np.eye(2)) @ rho @ np.kron(h, np.eye(2)).T)
entropy2 = entropy(np.kron(np.eye(2), h) @ rho @ np.kron(np.eye(2), h).T)
entropy3 = entropy(np.kron(h, h) @ rho @ np.kron(h, h).T)

print(f"Entropy Psi: {entropy0:.2f}")
print(f"Entropy (H I) Psi: {entropy1:.2f}")
print(f"Entropy (I H) Psi: {entropy2:.2f}")
print(f"Entropy (H H) Psi: {entropy3:.2f}")

Entropy Psi: -0.00
Entropy (H I) Psi: 0.69
Entropy (I H) Psi: 0.69
Entropy (H H) Psi: 1.39


Давайте еще раз выпишем, что получилось:

- $S(\ket{\Psi}) = 0$
- $S((\hat{H} \otimes \hat{I})\ket{\Psi}) = 0.69$
- $S((\hat{I} \otimes \hat{H})\ket{\Psi}) = 0.69$
- $S((\hat{H} \otimes \hat{H})\ket{\Psi}) = 1.39$

Таким образом, гейт Адамара увеличивает энтропию фон-Неймана нашей системы, именно поэтому большинство квантовых схем начинается именно с этого оператора. Часто даже звучит фраза "Сначала переведем нашу систему в запутанное состояние оператором Адамара". Только что мы увидели численное подтверждение этому выражению!

## Что мы узнали?

- Формализм матрицы плотности позволяет описывать составные системы (например, один кубит в многокубитной системе)
- Чаще всего в реальных экспериментах у нас ситуация "мы приготовили состояние, но точно не знаем какое" и волновая функция нам не подходит
- Что такое квантовая запутанность и как
- Состояние части запутанного состояния -- смешанное
- Отличие смешанного состояния от чистого можно охарактеризовать параметром типа энтропии